### PR TITLE
Add array as an alternative input + `code` option

### PR DESCRIPTION
**Pull request by gforge**
This PR does the following:

- Allows for `keyValue`/`keyCode` to come in array format
- Changes behavior to allow simultaneous input of all three options (**API change**) see discussion below &dagger;
- Adds the option of having `code` instead of `keyCode` (also marks `keyCode` as deprecated)
- Adds an example for toying around with `code`
- Clarifies the examples by visually boxing each example and adding code sections where appropriate
- Refactoring of the code so that `index.js` is more informative, decorators have a separate file + more
- Updated the README and also entered info on the missing section (much is copy-paste from MDN)

The PR includes all the updates in previous Flow PR (https://github.com/ayrton/react-key-handler/pull/147)

&dagger; The possibility of `keyValue` having priority is problematic in my opinion. If a user does not intend to use both then they probably wouldn't have specified them. If this truly should not be allowed then the constructor should throw an error. I can't see any use case where allowing all provided value hooks to trigger the event function.

### CODE REVIEW

File: demo/components/examples/decorators/key-handler.js [hunk 1]
```diff
@@ -22,8 +23,15 @@ function Demo({keyValue}: Props) {
...
+      </p>
+      <pre>
+        {'keyToggleHandler({keyEventName: KEYPRESS, keyValue: \'s\'})(Component)'}
+      </pre>
+    </ExampleBox>
   );
 }
 
-export default keyHandler({keyEventName: KEYPRESS, keyValue: 's'})(Demo);
+export default keyToggleHandler({keyEventName: KEYPRESS, keyValue: 's'})(Demo);
```
[the remark this answers is not in the document]
Comment (gforge) [reply demo/components/examples/decorators/key-handler.js]: Fixed in new example. I've also added the `react-syntax-highlighter` package and an h3-style to make the examples easier to read.

**Comment by gforge** [comment]
I would love your opinion on the file naming and structure. There are a few things that I've been thinking about:

- `key-handler-decorator`: Should probably just be `decorator`
- `utils.js`: This file is stuffed with functions. I personally prefer having a folder with an index file that merges all the functions.
- `keyHandler` and `keyToggleHandler`: These should perhaps be defined in the `index.js`

I've been using `rollup` for some of my libraries and it allows more `mjs` and other output formats that could be interesting optimizations: https://medium.com/@tarkus/how-to-build-and-publish-es6-modules-today-with-babel-and-rollup-4426d9c7ca71

File: README.md [hunk 2]
```diff
@@ -2,7 +2,7 @@
```
Comment (ayrton) [inline README.md]: These README changes are +💯

File: lib/index.js [hunk 3]
```diff
@@ -1 +1,6 @@
-module.exports = require('./key-handler');
+/* @flow */
+import KeyHandler from './key-handler';
+export {KEYDOWN, KEYPRESS, KEYUP} from './constants';
+export {default as keyHandleDecorator, keyToggleHandler, keyHandler} from './key-handle-decorator';
+
+export default KeyHandler;
```
Comment (ayrton) [inline lib/index.js]: Won't the end-user now have to use something like webpack? (vs. when using `module.exports`)

File: lib/key-handle-decorator.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+import React from 'react';
+
+import KeyHandler from './key-handler';
+import {type KeyConstants} from './constants';
+import {matchesKeyboardEvent, eventKey} from './utils';
+import { type rawOrArray } from './types';
+
+export type KeyhandleDecoratorState = {|
+ keyValue?: ?rawOrArray<string>,
```
Comment (ayrton) [inline lib/key-handle-decorator.js]: Wonder if there's a better name for `raw`, or is this something commonly used?

File: lib/key-handle-decorator.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      class KeyHandleDecorator extends React.Component<T, KeyhandleDecoratorState> {
+        state: KeyhandleDecoratorState = {
+          keyCode: null,
+          keyValue: null,
+          code: null,
+        }
+
+        render() {
+          return (
+            <div>
```
Comment (ayrton) [inline lib/key-handle-decorator.js]: We should probably do `React.Fragment` here, wdyt?

**Comment by ayrton** [comment]
@gforge 

> I would love your opinion on the file naming and structure. There are a few things that I've been thinking about:

Let's discuss this in an issue or separate PR

File: lib/index.js (hunk 3 above)
Comment (gforge) [reply lib/index.js]: I don't think so but if we go with rollup it's easy to deliver umd, es6, cjs and other packages

File: lib/utils.js [hunk 6]
```diff
@@ -100,15 +35,33 @@ function isContentEditable(element: HTMLElement): boolean {
  * Matches an event against a given keyboard key.
  */
 
-export function matchesKeyboardEvent(event: KeyboardEvent, {keyCode, keyValue}: KeyboardKey): boolean {
+function elmntOrArrayMatch<T>(a: T | T[], b: T) {
```
[the remark this answers is not in the document]
Comment (gforge) [reply lib/utils.js]: This is just a helper to the original matchesKeyboardEvent, not sure I understand

File: package.json [hunk 7]
```diff
@@ -68,6 +68,7 @@
   },
   "dependencies": {
     "exenv": "^1.2.0",
-    "prop-types": "^15.6.2"
+    "prop-types": "^15.6.2",
+    "react-syntax-highlighter": "^8.0.1"
```
[the remark this answers is not in the document]
Comment (gforge) [reply package.json]: Will change, writing on the phone.

File: test/support/helper.js [hunk 8]
```diff
@@ -3,8 +3,6 @@ import chaiEnzyme from 'chai-enzyme';
 import { configure } from 'enzyme';
 import Adapter from 'enzyme-adapter-react-16';
 import { JSDOM } from 'jsdom';
-import path from 'path';
-import requirePaths from 'app-module-path';
```
[the remark this answers is not in the document]
Comment (gforge) [reply test/support/helper.js]: The problem was that my IDE complained that it couldn't find the utils package.

I think though that tests should be super easy to understand. The app-module-path requires that you first check the setup and also understand the package. 

As a side note, I wasn't sure that I would get this great interaction. Much of the changes were made to adapt to my own coding style as I was certain to post my own scoped version of the package. There is an abundance of abandoned packages out there, given how annoying updating all the dependencies are, it's quite understandable :grin:

**Comment by gforge** [comment]
Updated the PR with the fixes

File: lib/index.js (hunk 3 above)
Comment (ayrton) [reply lib/index.js]: Rollup definitely wasn't as popular as it is now when this package was initially released, seems most libraries moved to it, and sounds like we should too

File: test/support/helper.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (ayrton) [reply test/support/helper.js]: Let's remove the `app-module-path` dependency

File: lib/key-handle-decorator.js (hunk 4 above)
Comment (ayrton) [reply lib/key-handle-decorator.js]: I like `primitiveOrArary`

File: lib/utils.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (ayrton) [reply lib/utils.js]: Suggesting a function rename, to make it more in lines of the "convention" of the other function `matchesKeyBoardEvent`

File: lib/key-handle-decorator.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      class KeyHandleDecorator extends React.Component<T, KeyhandleDecoratorState> {
+        state: KeyhandleDecoratorState = {
+          keyCode: null,
+          keyValue: null,
+          code: null,
+        }
+
+        render() {
+          return (
+            <React.Fragment>
```
Comment (ayrton) [inline lib/key-handle-decorator.js]: We'll have to update our peerDependencies in package.json:

``` json
{
  "peerDependencies": {
    "react": ^16.2.0-0"
  }
}
```

Comment (gforge) [reply lib/key-handle-decorator.js]: Fixed

File: test/support/helper.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (gforge) [reply test/support/helper.js]: Fixed

File: lib/utils.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (gforge) [reply lib/utils.js]: Excellent suggestion, changed

File: lib/key-handle-decorator.js (hunk 4 above)
Comment (gforge) [reply lib/key-handle-decorator.js]: Fixed

File: lib/index.js (hunk 3 above)
Comment (gforge) [reply lib/index.js]: Sure, I will prepare a PR for it once we're done with this. The docs make it sound like it should be super easy but then you start looking into all the options and its a :rabbit2: :hole: I have a good setup in a similar project although in that one I used Typescript so it needs some tweaking before we can deploy it.

I'm not a great fan but I think webpack-dev-server is super convenient for the demo and therefore we may want to keep it in place.

**Comment by ayrton** [comment]
@gforge can you rebase, would love to get this in for 1.2.0

**Comment by gforge** [comment]
Ok, rebased from master

File: README.md [hunk 10]
```diff
@@ -124,32 +128,43 @@ The prop types of the `KeyHandler` component are:
...
 This library does not handle key events for form elements such as `<input />` and `<textarea />`.
 
-React does a fine job supporting these already via [keyboard events](https://facebook.github.io/react/docs/events.html#keyboard-events).
+React does a fine job supporting these already via [keyboard events](https://reactjs.org/docs/events.html#keyboard-events).
 
 [Examples](demo/components/examples/input/)
 
 ## Key event names
 
-TODO: explain the differences between the different key events.
```
Comment (ayrton) [inline README.md]: Can you leave the above and add a new section for the new documentation you've added: 

```md
## `keyValue'`, `code` and `keyCode`
```

We'll also need to update the TOC because the heading changed

File: README.md [hunk 11]
```diff
@@ -124,32 +131,47 @@ The prop types of the `KeyHandler` component are:
...
 ## Key event names
 
 TODO: explain the differences between the different key events.
 
-## `keyValue`, `keyCode`
+## `keyValue`, `code` and `keyCode`
+
+The three available key events are
+
+- `key`      This corresponds to the true value. This is the value of the key pressed
```
Comment (ayrton) [inline README.md]: this should be `keyValue` vs. `key`

File: lib/key-handler.js [hunk 12]
```diff
@@ -34,8 +46,11 @@ export default class KeyHandler extends React.Component<Props> {
     super(props);
 
     /* eslint-disable no-console */
+    if (props.keyCode) {
+      console.warn('The `keyCode` is deprecated, use `code` instead');
+    }
 
-    if (!props.keyValue && !props.keyCode) {
+    if (!props.keyValue && !props.keyCode && !props.code) {
       console.error('Warning: Failed propType: Missing prop `keyValue` or `keyCode` for `KeyHandler`.');
```
Comment (ayrton) [inline lib/key-handler.js]: Can you update the error message to include `code`

File: lib/key-handler.js [hunk 13]
```diff
@@ -34,8 +46,11 @@ export default class KeyHandler extends React.Component<Props> {
     super(props);
 
     /* eslint-disable no-console */
+    if (props.keyCode) {
+      console.warn('The `keyCode` is deprecated, use `code` instead');
```
Comment (ayrton) [inline lib/key-handler.js]:
```js
console.warn('Warning: Deprecated propType: `keyCode` is deprecated in favour of `code` for `KeyHandler`.');
```

File: lib/key-handle-decorator.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+              <KeyHandler
+                keyValue={keyValue}
+                keyCode={keyCode}
+                code={code}
+                keyEventName={keyEventName}
+                onKeyHandle={this.handleKey}
+              />
+              <Component {...otherProps} {...this.state} />
```
Comment (ayrton) [inline lib/key-handle-decorator.js]: this should be `<Component {...this.props} {...this.state} />` otherwise the decorated component will not receive their original props

File: lib/key-handler.js [hunk 15]
```diff
@@ -4,20 +4,32 @@ import React from 'react';
...
+import {KEYDOWN, KEYPRESS, KEYUP, type KeyConstants} from './constants';
+import {type primitiveOrArray} from './types';
+import {isInput, matchesKeyboardEvent} from './utils';
 
 type Props = {|
-  +keyValue?: string,
-  +keyCode?: number,
-  +keyEventName: KEYDOWN | KEYPRESS | KEYUP,
-  +onKeyHandle?: (event: KeyboardEvent) => void,
+  +keyValue?: ?primitiveOrArray<string>,
```
Comment (ayrton) [inline lib/key-handler.js]: What are the `+` prefixes for?

File: test/components/key-handler.test.js [hunk 16]
```diff
@@ -71,7 +82,10 @@ describe('KeyHandler', () => {
     expect(handler.calledOnce).to.equal(true);
   });
 
-  it('prioritizes key value over code', () => {
+  // This make little sense - a user will not understand that there is a priority
+  // the user wil most likely expect that the function gets called if any
+  // of the matching criteria match.
+  it.skip('prioritizes key value over code', () => {
```
Comment (ayrton) [inline test/components/key-handler.test.js]: If this makes no sense we should `invariant` in the constructor

File: test/components/key-handler.test.js [hunk 17]
```diff
@@ -19,13 +19,24 @@ describe('KeyHandler', () => {
 
   it('handles key up events when key value match', () => {
     const handler = sinon.spy();
-    mount(<KeyHandler keyValue={ARROW_LEFT} onKeyHandle={handler} />);
+    mount(<KeyHandler keyValue={'s'} onKeyHandle={handler} />);
 
-    triggerKeyEvent(KEYUP, undefined, ARROW_LEFT);
+    triggerKeyEvent(KEYUP, undefined, 's');
```
Comment (ayrton) [inline test/components/key-handler.test.js]: why were these changed?

**Comment by ayrton** [comment]
Why the 5bcdd9a change? Since it's no official release I'd just revert that commit

File: lib/key-handler.js (hunk 15 above)
Comment (ayrton) [reply lib/key-handler.js]: TIL they indicate the props are readonly, see flow docs https://flow.org/en/docs/types/interfaces/#toc-interface-property-variance-read-only-and-write-only

Comment (gforge) [reply lib/key-handler.js]: Sorry, wrote the answers yesterday but there was a "pending" set to the answer. TIL - you need to submit pending comments at the top of the page :grimacing:

File: test/components/key-handler.test.js (hunk 16 above)
Comment (ayrton) [reply test/components/key-handler.test.js]: It's an [npm package](https://www.npmjs.com/package/invariant)

[the remark this answers is not in the document]
Comment (ayrton) [reply test/components/key-handler.test.js]: yep I'm okay with this braking change, dropping the test makes sense once we throw the invariant

**Comment by gforge** [comment]
Thanks for merging! Sorry, didn't check babylon & parcel compatibility. Also, I like the prettier code style change :smile:

**Comment by ayrton** [comment]
@gforge no worries, you did an epic job here! Just released v1.2.0-beta.3 so feel free to test it out

**Comment by gforge** [comment]
Thanks. I've tested it and it seems to work fine.

I think you should do a major release since the handling of simultaneous keycodes has changed
